### PR TITLE
Pytorch version

### PR DESCRIPTION
**Pull request by wronaq**

### CODE REVIEW

File: color_generator/datasets/dataloaders.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import numpy as np
+from torch.utils.data import DataLoader, Subset
+from torch.utils.data.sampler import SubsetRandomSampler
+
+
+class DataLoaders:
```
Comment (blawok) [inline color_generator/datasets/dataloaders.py:6]: Nicely done but it could be omitted by using datasets (https://huggingface.co/docs/datasets/quicktour.html#quick-tour)

File: color_generator/models/base.py [hunk 2]
```diff
@@ -1,71 +1,121 @@
...
+    @property
+    def weights_filename(self):
+        p = Path(__file__).resolve().parents[2] / "weights"
+        p.mkdir(parents=True, exist_ok=True)
+        return str(p / f"{self.name}_weights.pt")
+
+    def fit(self, epochs=10):
+
+        criterion = self.criterion()
+        cs = nn.CosineSimilarity(dim=1)
```
Comment (blawok) [inline color_generator/models/base.py:33]: Great idea to use cosine similarity as an additional metric

File: color_generator/models/base.py [hunk 3]
```diff
@@ -1,71 +1,121 @@
...
-            dataset,
-            epochs=epochs,
-            callbacks=callbacks,
-            # validation_data=test_sequence,
-            use_multiprocessing=False,
-            workers=1,
-            shuffle=True,
+        self.name = f"{self.__class__.__name__}_{self.dataset.__class__.__name__}_{self.network.__class__.__name__}"
+        self._early_stopping = EarlyStopping(
+            patience=1, verbose=True, delta=0.001, path="early_stopping_checkpoint.pt"
```
Comment (blawok) [inline color_generator/models/base.py:21]: Early stopping hyperparameters could be parameterized instead of hardcoding them

File: color_generator/models/base.py [hunk 4]
```diff
@@ -1,71 +1,121 @@
...
+
+        for epoch in range(epochs):
+            self.network.train()
+            running_loss = 0.0
+            running_cs = 0.0
+            for i, batch in enumerate(self._dataloaders.train_loader):
+                # forward and backward propagation
+                input_ids = batch["input_ids"].to(self.device)
+                attention_mask = batch["attention_mask"].to(self.device)
+                targets = batch["target"].to(self.device)
```
Comment (blawok) [inline color_generator/models/base.py:43]: This could be done using:
```
batch = {k: v.to(device) for k, v in batch.items()}
outputs = model(**batch)
```

File: color_generator/models/color_model.py [hunk 5]
```diff
@@ -1,33 +1,30 @@
...
-            'input_ids': convert_to_tensor([input_text['input_ids']], np.int32),
-            'attention_mask': convert_to_tensor([input_text['attention_mask']], np.int32)
-        })
-
-        return predictions
+    def __init__(self, dataloaders, network_fn, device):
+        super().__init__(dataloaders, network_fn, device)
+
+    def predict_on_text(self, input_text):
+        tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")
```
Comment (blawok) [inline color_generator/models/color_model.py:19]: Same as in color_datasets you could use AutoTokenizer and parameterize model name

File: color_generator/models/color_model.py [hunk 6]
```diff
@@ -1,33 +1,30 @@
...
+        input_text = tokenizer(input_text, truncation=True, return_tensors="pt")
```
Comment (blawok) [inline color_generator/models/color_model.py:20]: If you would limit the sequence length, you would have to remember to change this - maybe it is worth to extract the tokenization process into separate module?

File: color_generator/models/color_model.py [hunk 7]
```diff
@@ -1,33 +1,30 @@
...
+        self.network.eval()
+        predictions = self.network(
+            input_text["input_ids"], input_text["attention_mask"]
+        ).squeeze()
+
+        return self._unnorm(predictions)
+
+    @staticmethod
+    def _unnorm(predictions):
```
Comment (blawok) [inline color_generator/models/color_model.py:29]: Maybe change the method name to `_decode(predictions)`?

File: color_generator/networks/distilbert.py [hunk 8]
```diff
@@ -1,26 +1,27 @@
-"""Define DistilBERT network function."""
-from typing import Tuple
+import torch.nn as nn
+from transformers import DistilBertModel, DistilBertConfig
```
Comment (blawok) [inline color_generator/networks/distilbert.py:2]: This could also be done with AutoModel and AutoConfig (https://huggingface.co/transformers/model_doc/auto.html#automodel) to allow different architectures that have the same input and output

**Review by blawok — APPROVED** [review]
Great job, repo looks much cleaner now and it will allow us to continue work on this project.
It would be nice to add some unit tests next time :)

I added some comments but none of them requires any changes before merging - we could think about those changes in the future.
